services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/immutable/ImmutableIntrospector.java

```diff
@@ -0,0 +1,226 @@
+
+        if (getter.getParameterCount() != 0) {
+            throw generateExceptionForMethod(getter, "A method was found on the immutable class that does not appear " +
+                "to be a valid getter due to it having one or more parameters.");
+        }
+    }
+
+    private List<Method> filterAndCollectGetterMethods(Method[] rawMethods) {
+        return Arrays.stream(rawMethods)
+                     .filter(ImmutableIntrospector::isMappableMethod)
+                     .collect(Collectors.toList());
+    }
+
+    private Map<String, Method> filterAndIndexBuilderMethods(Method[] rawMethods) {
+        return Arrays.stream(rawMethods)
+                     .filter(ImmutableIntrospector::isMappableMethod)
+                     .collect(Collectors.toMap(this::normalizeSetterName, m -> m));
+    }
+
+    private String normalizeSetterName(Method setter) {
+        String setterName = setter.getName();
+
+        if (setterName.length() > 3
+            && Character.isUpperCase(setterName.charAt(3))
+            && setterName.startsWith(SET_PREFIX)) {
+
+            return Character.toLowerCase(setterName.charAt(3)) + setterName.substring(4);
+        }
+
+        return setterName;
+    }
+
+    private String normalizeGetterName(Method getter) {
+        String getterName = getter.getName();
+
+        if (getterName.length() > 2
+            && Character.isUpperCase(getterName.charAt(2))
+            && getterName.startsWith(IS_PREFIX)
+            && isMethodBoolean(getter)) {
+
+            return Character.toLowerCase(getterName.charAt(2)) + getterName.substring(3);
+        }
+
+        if (getterName.length() > 3
+            && Character.isUpperCase(getterName.charAt(3))
+            && getterName.startsWith(GET_PREFIX)) {
+
+            return Character.toLowerCase(getterName.charAt(3)) + getterName.substring(4);
+        }
+
+        return getterName;
+    }
+
+    private boolean isMethodBoolean(Method method) {
+        return method.getReturnType() == boolean.class || method.getReturnType() == Boolean.class;
+    }
+
+    private Optional<Method> extractBuildMethod(Map<String, Method> indexedBuilderMethods, Class<?> immutableClass) {
+        Method buildMethod = indexedBuilderMethods.get(BUILD_METHOD);
+
+        if (buildMethod == null
+            || buildMethod.getParameterCount() != 0
+            || !immutableClass.equals(buildMethod.getReturnType())) {
+
+            return Optional.empty();
+        }
+
+        indexedBuilderMethods.remove(BUILD_METHOD);
+        return Optional.of(buildMethod);
+    }
+
+    private Optional<Method> extractSetterMethod(String propertyName,
+                                                 Map<String, Method> indexedBuilderMethods,
+                                                 Method getterMethod,
+                                                 Class<?> builderClass) {
+        Method setterMethod = indexedBuilderMethods.get(propertyName);
+
+        if (setterMethod == null
```

Comment:
Can we break up the if statement a bit, like creating variables with meaningful names? seems a bit hard to read with so many predicates.

Reply:
Okay. +1